src/core/include/openvino/op/constant.hpp

```diff
@@ -444,29 +432,49 @@ class OPENVINO_API Constant : public Op {
 #    pragma warning(disable : 4018)
 #    pragma warning(disable : 4804)
 #endif
-            if (!std::is_same<OUT_T, IN_T>::value) {
-                OPENVINO_ASSERT(!std::numeric_limits<IN_T>::is_signed || std::numeric_limits<OUT_T>::lowest() <= c,
-                                "Cannot cast vector from ",
-                                Type,
-                                " constant to ",
-                                element::from<OUT_T>(),
-                                ". Some values are outside the range. Example: ",
-                                c);
-                OPENVINO_ASSERT(std::numeric_limits<OUT_T>::max() >= c,
-                                "Cannot cast vector from ",
-                                Type,
-                                " constant to ",
-                                element::from<OUT_T>(),
-                                ". Some values are outside the range. Example: ",
-                                c);
-            }
+    template <class U,
+              class ConstantT,
+              typename std::enable_if<!std::is_unsigned<ConstantT>::value &&
+                                      !std::is_same<U, ConstantT>::value>::type* = nullptr>
+    static bool in_type_range(const ConstantT v) {
+        return std::numeric_limits<U>::lowest() <= v && v <= std::numeric_limits<U>::max();
+    }
+
+    template <class U,
+              class ConstantT,
+              typename std::enable_if<std::is_unsigned<ConstantT>::value && !std::is_same<U, ConstantT>::value>::type* =
+                  nullptr>
+    static bool in_type_range(const ConstantT v) {
+        return v <= std::numeric_limits<U>::max();
+    }
 #if defined(__clang__)
 #    pragma clang diagnostic pop
 #elif defined(__GNUC__)
 #    pragma GCC diagnostic pop
 #elif defined(_MSC_VER)
 #    pragma warning(pop)
 #endif
+
+    template <class U, class ConstantT, typename std::enable_if<std::is_same<U, ConstantT>::value>::type* = nullptr>
+    static constexpr bool in_type_range(const ConstantT) {
+        return true;
+    }
+
+    template <
+        element::Type_t Type,
+        class OUT_T,
+        typename std::enable_if<Type != element::string && !std::is_same<OUT_T, std::string>::value>::type* = nullptr>
+    void cast_vector(std::vector<OUT_T>& output_vector, size_t num_elements) const {
+        using InputT = ov::fundamental_type_for<Type>;
+        auto first = get_data_ptr<InputT>();
+        std::transform(first, first + num_elements, std::back_inserter(output_vector), [](const InputT c) {
+            OPENVINO_ASSERT(in_type_range<OUT_T>(c),
```

Comment:
Do we really need this check? It will fail when we are doing potentially inaccurate casts, like bf16 -> f16. Besides, it is not a performance friendly -- we are checking it in a regular lool over all elements. And there is no clear workaround when it really fails -- how to efficiently apply some saturation rule to this process of conversion? Old logic of cast_vector didn't imply that it is an efficient way of the cast -- it checks each value.